user-manual/src/config/kdl.md

```diff
@@ -255,6 +264,121 @@ Filters at this stage are the earliest. Currently supported filters:
+
+The bucket is refilled at a configurable rate, specified by `refill-rate-ms`, and adds a configurable
+number of tokens specified by `refill-qty`. The number of tokens in the bucket will never exceed the
+initial `tokens-per-bucket` number.
+
+Once a refill occurs, requests may become ready if a token becomes available.
+
+[Leaky Bucket]: https://en.wikipedia.org/wiki/Leaky_bucket
+
+##### How many buckets?
+
+Some rules require many buckets. For example, rules based on the source IP address will create a bucket
+for each unique IP address of downstream users.
+
+However, each of these buckets require space to contain the metadata, and to avoid unbounded growth,
+we allow for a configurable `max-buckets` number, which serves to influence the total memory required
+for storing buckets. This uses an [Adaptive Replacement Cache]
+to allow for concurrent access to these buckets, as well as the ability to automatically buckets that
+are not actively being used (somewhat similar to an LRU or "Least Recently Used" cache).
+
+[Adaptive Replacement Cache]: https://docs.rs/concread/latest/concread/arcache/index.html
+
+There is a trade off here: The larger `max-buckets` is, the longer that River can "remember" a bucket
+for a given factor, such as specific IP addresses. However, it also requires more resident memory to
+retain this information.
+
+If `max-buckets` is set too low, then buckets will be "evicted" from the cache, meaning that subsequent
+requests matching that bucket will require the creation of a new bucket (with a full set of tokens),
+potentially defeating the objective of accurate rate limiting.
+
+##### Gotta claim 'em all
+
+When multiple rules apply to a single request, for example rules based on both source IP address,
+and the URI path, then a request must claim ALL applicable tokens before proceeding. If a given IP
+address is making it's first request, but to a URI that that has an empty bucket, it will immediately
+obtain the IP address token, but be forced to wait until the URI token has been claimed
+
+##### Kinds of Rules
+
+Currently two kinds of rules are supported:
+
+* `kind="source-ip"` - this tracks the IP address of the requestor. A unique bucket will be created for
+  the IPv4 or IPv6 address of the requestor.
+* `kind="uri" pattern="REGEX"` - This tracks the URI path of the request, such as `static/images/example.jpg`
+    * If the request's URI path matches the provided `REGEX`, the full URI path will be assigned to a given
+      bucket
+    * For example, if the regex `static/.*` was provided:
+        * `index.html` would not match this rule, and would not require obtaining a token
+        * `static/images/example.jpg` would match this rule, and would require obtaining a token
+        * `static/styles/example.css` would also match this rule, and would require obtaining a token
+        * Note that `static/images/example.jpg` and `static/styles/example.css` would each have a UNIQUE
```

Comment:
I think (probably as an eventual feature) it would be very nice to rate-limit groups of endpoints together -- we do exactly this with our rate-limiter, and without reading docs I would guess a single path rule would have a single bucket used for all matching requests. I think we should be careful that we aren't painting ourselves into a corner here.

Reply:
This is a reasonable request, and I did go back and forth on the potential behavior! I could see it being useful to have a "single bucket rule" instead of a "multi bucket rule", like:

* one kind of rule that makes two buckets for `static/.*` for `static/images/example.jpg` and `static/styles/example.css`
* one kind of rule that makes exactly one bucket for `static/.*`, so `static/images/example.jpg` and `static/styles/example.css` take from the same bucket

I'll see if this is easy to specify in this PR, or open an issue to track future filter kinds.

Reply:
Nice! Yep, this matches my expectations.